File: platform/test/integration/storage/s3-cached-key/test_case.py
#!/usr/bin/env python3
"""
Integration test for cache_json_awss3_storage — S3 JSON Cache workbench.

Mirrors the webapp workbench operations:
  1. Resolves config via get_client_factory_from_app_config (three-tier).
  2. Creates an async S3 client + JsonS3Storage instance.
  3. SAVE  — write a JSON entry with TTL.
  4. LOAD  — read it back, verify data round-trips.
  5. EXISTS — confirm the key is present.
  6. LIST  — list keys with prefix, confirm the test key appears.
  7. DELETE — remove the key.
  8. LOAD (miss) — confirm load returns None after delete.
  9. Cleanup: close storage + client.
"""

from __future__ import annotations

import argparse
import asyncio
import json
import os
import re
import sys
import time
from pathlib import Path

_HERE = Path(__file__).resolve().parent
_PLATFORM = _HERE
while _PLATFORM.name != "platform" and _PLATFORM != _PLATFORM.parent:
    _PLATFORM = _PLATFORM.parent


def _inject_paths():
    """Add polyglot/packages_py source dirs to sys.path from platform pyproject.toml."""
    toml = _PLATFORM / "pyproject.toml"
    if not toml.exists():
        return
    for line in toml.read_text().splitlines():
        m = re.search(r'path\s*=\s*"([^"]+)"', line)
        if not m:
            continue
        pkg_path = (_PLATFORM / m.group(1)).resolve()
        src = pkg_path / "src"
        sys.path.insert(0, str(src if src.is_dir() else pkg_path))


_inject_paths()

from app_yaml_load import load_app_yaml_config
from cache_json_awss3_storage.config_bridge import get_client_factory_from_app_config
from cache_json_awss3_storage import create_storage, ClientAsync


# ---------------------------------------------------------------------------
# Test data
# ---------------------------------------------------------------------------

TEST_KEY = f"integ-py-{int(time.time() * 1000)}"
TEST_DATA = {
    "source": "test_case.py",
    "id": TEST_KEY,
    "items": [1, 2, 3],
    "nested": {"ok": True},
}
TEST_TTL = 600
KEY_PREFIX = "jss3:"


# ---------------------------------------------------------------------------
# Run workbench-style operations
# ---------------------------------------------------------------------------

DEFAULT_TEST_BUCKET = "figma-component-inspector"


async def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--app-env", default="dev")
    parser.add_argument("--config-dir", required=True)
    parser.add_argument("--bucket", default="")
    args = parser.parse_args()

    app_cfg = load_app_yaml_config(config_dir=args.config_dir, app_env=args.app_env).config
    yaml_s3 = app_cfg.get_nested("storage", "s3", default={}) or {}

    # Resolve bucket: CLI arg → env → YAML → default test bucket
    bucket_override = args.bucket or os.environ.get("AWS_S3_BUCKET", "")
    overrides = {}
    if bucket_override:
        overrides["bucket_name"] = bucket_override

    cfg = get_client_factory_from_app_config(yaml_s3, **overrides)

    # If config resolution left bucket empty, use the default test bucket
    if not cfg.bucket_name:
        cfg.bucket_name = bucket_override or DEFAULT_TEST_BUCKET

    results: dict = {
        "service": "s3-cached-key",
        "mode": "storage-workbench",
        "python_package": "cache_json_awss3_storage",
        "app_yaml_initialized": True,
        "config": {
            "bucket_name": cfg.bucket_name,
            "region_name": cfg.region_name,
            "endpoint_url": cfg.endpoint_url or "(default)",
            "key_prefix": KEY_PREFIX,
            "ttl": TEST_TTL,
        },
        "steps": {},
    }

    try:
        async with ClientAsync(cfg) as s3_client:
            # Step 1: Create storage
            storage = create_storage(
                s3_client,
                bucket_name=cfg.bucket_name,
                key_prefix=KEY_PREFIX,
                ttl=TEST_TTL,
                region=cfg.region_name,
            )
            results["steps"]["create_storage"] = {"ok": True}

            # Step 2: SAVE
            save_start = time.monotonic()
            saved_key = await storage.save(TEST_KEY, TEST_DATA, ttl=TEST_TTL)
            save_latency = round((time.monotonic() - save_start) * 1000)
            results["steps"]["save"] = {
                "ok": saved_key == TEST_KEY,
                "key": saved_key,
                "s3_key": f"{KEY_PREFIX}{saved_key}",
                "ttl": TEST_TTL,
                "latency_ms": save_latency,
            }

            # Step 3: LOAD
            load_start = time.monotonic()
            loaded = await storage.load(TEST_KEY)
            load_latency = round((time.monotonic() - load_start) * 1000)
            data_matches = loaded is not None and loaded == TEST_DATA
            results["steps"]["load"] = {
                "ok": data_matches,
                "key": TEST_KEY,
                "found": loaded is not None,
                "data_matches": data_matches,
                "latency_ms": load_latency,
            }

            # Step 4: EXISTS
            exists_start = time.monotonic()
            does_exist = await storage.exists(TEST_KEY)
            exists_latency = round((time.monotonic() - exists_start) * 1000)
            results["steps"]["exists"] = {
                "ok": does_exist is True,
                "key": TEST_KEY,
                "exists": does_exist,
                "latency_ms": exists_latency,
            }

            # Step 5: LIST
            list_start = time.monotonic()
            keys = await storage.list_keys()
            list_latency = round((time.monotonic() - list_start) * 1000)
            test_key_found = TEST_KEY in keys
            results["steps"]["list"] = {
                "ok": test_key_found,
                "key_prefix": KEY_PREFIX,
                "count": len(keys),
                "test_key_found": test_key_found,
                "latency_ms": list_latency,
            }

            # Step 6: DELETE
            delete_start = time.monotonic()
            deleted = await storage.delete(TEST_KEY)
            delete_latency = round((time.monotonic() - delete_start) * 1000)
            results["steps"]["delete"] = {
                "ok": deleted is True,
                "key": TEST_KEY,
                "deleted": deleted,
                "latency_ms": delete_latency,
            }

            # Step 7: LOAD after delete (should be None)
            load_miss_start = time.monotonic()
            loaded_after = await storage.load(TEST_KEY)
            load_miss_latency = round((time.monotonic() - load_miss_start) * 1000)
            results["steps"]["load_after_delete"] = {
                "ok": loaded_after is None,
                "key": TEST_KEY,
                "found": loaded_after is not None,
                "latency_ms": load_miss_latency,
            }

            await storage.close()

        # Overall
        all_ok = all(step.get("ok") for step in results["steps"].values())
        results["overall"] = "PASS" if all_ok else "FAIL"

    except Exception as exc:
        results["error"] = {"name": type(exc).__name__, "message": str(exc)}
        results["overall"] = "FAIL"

    print(json.dumps(results, indent=2))
    return 0


if __name__ == "__main__":
    raise SystemExit(asyncio.run(main()))

File: platform/test/integration/storage/s3-cached-key/test_case.mjs
#!/usr/bin/env node

/**
 * Integration test for cache_json_awss3_storage — S3 JSON Cache workbench.
 *
 * Mirrors the webapp workbench operations:
 *   1. Resolves config via getClientFactoryFromAppConfig (three-tier).
 *   2. Creates an async S3 client + JsonS3Storage instance.
 *   3. SAVE  — write a JSON entry with TTL.
 *   4. LOAD  — read it back, verify data round-trips.
 *   5. EXISTS — confirm the key is present.
 *   6. LIST  — list keys with prefix, confirm the test key appears.
 *   7. DELETE — remove the key.
 *   8. LOAD (miss) — confirm load returns null after delete.
 *   9. Cleanup: close storage + destroy client.
 */

import path from 'node:path';
import { pathToFileURL } from 'node:url';

const args = process.argv.slice(2);
const readArg = (name, fallback) => {
  const idx = args.indexOf(name);
  return idx >= 0 ? args[idx + 1] : fallback;
};

const appEnv = readArg('--app-env', process.env.APP_ENV || 'dev');
const configDir = readArg('--config-dir', process.env.CONFIG_DIR);
if (!configDir) {
  console.error('Missing --config-dir');
  process.exit(2);
}

const platformRoot = path.resolve(configDir, '..', '..');
const importFile = async (absPath) => import(pathToFileURL(absPath).href);

const appYamlLoadMod = await importFile(path.join(platformRoot, 'polyglot/app_yaml_load/mjs/dist/index.js'));
const cacheMod = await importFile(path.join(platformRoot, 'polyglot/cache_json_awss3_storage/mjs/dist/index.js'));

const { config: appConfig } = await appYamlLoadMod.loadAppYamlConfig({ configDir, appEnv });
const yamlS3 = appConfig.getNested(['storage', 's3'], {}) || {};

// Resolve bucket: CLI arg → YAML → env → default test bucket
const bucketOverride = readArg('--bucket', process.env.AWS_S3_BUCKET || '');
const DEFAULT_TEST_BUCKET = 'figma-component-inspector';

const cfg = cacheMod.getClientFactoryFromAppConfig(
  yamlS3,
  bucketOverride ? { bucketName: bucketOverride } : undefined,
);

// If config resolution left bucket empty, use the default test bucket
if (!cfg.bucketName) {
  cfg.bucketName = bucketOverride || DEFAULT_TEST_BUCKET;
}

// ---------------------------------------------------------------------------
// Test data
// ---------------------------------------------------------------------------

const TEST_KEY = `integ-mjs-${Date.now()}`;
const TEST_DATA = {
  source: 'test_case.mjs',
  id: TEST_KEY,
  items: [1, 2, 3],
  nested: { ok: true },
};
const TEST_TTL = 600;
const KEY_PREFIX = 'jss3:';

// ---------------------------------------------------------------------------
// Run workbench-style operations
// ---------------------------------------------------------------------------

const results = {
  service: 's3-cached-key',
  mode: 'storage-workbench',
  mjs_package: 'cache_json_awss3_storage',
  app_yaml_initialized: true,
  config: {
    bucketName: cfg.bucketName,
    regionName: cfg.regionName,
    endpointUrl: cfg.endpointUrl || '(default)',
    keyPrefix: KEY_PREFIX,
    ttl: TEST_TTL,
  },
  steps: {},
};

const handle = cacheMod.createAsyncClient(cfg);
let storage = null;

try {
  // Step 1: Create storage
  storage = cacheMod.createStorage({
    s3Client: handle.client,
    bucketName: cfg.bucketName,
    keyPrefix: KEY_PREFIX,
    ttl: TEST_TTL,
    region: cfg.regionName,
  });
  results.steps.create_storage = { ok: true };

  // Step 2: SAVE
  const saveStart = performance.now();
  const savedKey = await storage.save(TEST_KEY, TEST_DATA, { ttl: TEST_TTL });
  const saveLatency = Math.round(performance.now() - saveStart);
  results.steps.save = {
    ok: savedKey === TEST_KEY,
    key: savedKey,
    s3_key: `${KEY_PREFIX}${savedKey}`,
    ttl: TEST_TTL,
    latency_ms: saveLatency,
  };

  // Step 3: LOAD
  const loadStart = performance.now();
  const loaded = await storage.load(TEST_KEY);
  const loadLatency = Math.round(performance.now() - loadStart);
  const dataMatches = loaded !== null && JSON.stringify(loaded) === JSON.stringify(TEST_DATA);
  results.steps.load = {
    ok: dataMatches,
    key: TEST_KEY,
    found: loaded !== null,
    data_matches: dataMatches,
    latency_ms: loadLatency,
  };

  // Step 4: EXISTS
  const existsStart = performance.now();
  const doesExist = await storage.exists(TEST_KEY);
  const existsLatency = Math.round(performance.now() - existsStart);
  results.steps.exists = {
    ok: doesExist === true,
    key: TEST_KEY,
    exists: doesExist,
    latency_ms: existsLatency,
  };

  // Step 5: LIST
  const listStart = performance.now();
  const keys = await storage.listKeys();
  const listLatency = Math.round(performance.now() - listStart);
  const testKeyInList = keys.includes(TEST_KEY);
  results.steps.list = {
    ok: testKeyInList,
    key_prefix: KEY_PREFIX,
    count: keys.length,
    test_key_found: testKeyInList,
    latency_ms: listLatency,
  };

  // Step 6: DELETE
  const deleteStart = performance.now();
  const deleted = await storage.delete(TEST_KEY);
  const deleteLatency = Math.round(performance.now() - deleteStart);
  results.steps.delete = {
    ok: deleted === true,
    key: TEST_KEY,
    deleted,
    latency_ms: deleteLatency,
  };

  // Step 7: LOAD after delete (should be null)
  const loadMissStart = performance.now();
  const loadedAfterDelete = await storage.load(TEST_KEY);
  const loadMissLatency = Math.round(performance.now() - loadMissStart);
  results.steps.load_after_delete = {
    ok: loadedAfterDelete === null,
    key: TEST_KEY,
    found: loadedAfterDelete !== null,
    latency_ms: loadMissLatency,
  };

  // Overall
  const allOk = Object.values(results.steps).every((s) => s.ok);
  results.overall = allOk ? 'PASS' : 'FAIL';
} catch (err) {
  results.error = { name: err.name, message: err.message };
  results.overall = 'FAIL';
} finally {
  if (storage) {
    try { await storage.close(); } catch (_) { /* ignore */ }
  }
  try { handle.destroy(); } catch (_) { /* ignore */ }
}

console.log(JSON.stringify(results, null, 2));
